### ISSUE TITLE
Serialization hangs with specific HashMap values

### ISSUE THREAD

**Issue by pkwagner** [rkyv/rkyv#547]
Hi,

I've encountered a weird bug where serializing a HashMap with certain values causes the execution to hang indefinitely. This issue occurs on the latest version of rkyv (0.8.0) and has been observed across different platforms (ARM/macOS, ARM/Linux, and x86/Linux).

I've attached a minimal reproducible example below. When removing some of the entries from the HashMap, the serialization works correctly.

```rust
#[derive(Hash, PartialEq, Eq, rkyv::Archive, rkyv::Serialize, rkyv::Deserialize, Debug)]
#[rkyv(derive(Hash), derive(PartialEq), derive(Eq))]
struct Key(u8, u8);

fn main() {
    println!("Hello world!");

    let mut map = std::collections::HashMap::new();
    map.insert(Key(13, 2), 1337);
    map.insert(Key(5, 7), 1337);
    map.insert(Key(14, 5), 1337);
    map.insert(Key(23, 8), 1337);
    map.insert(Key(5, 5), 1337);
    map.insert(Key(24, 4), 1337);
    map.insert(Key(28, 1), 1337);
    map.insert(Key(29, 1), 1337);
    map.insert(Key(13, 3), 1337);
    map.insert(Key(30, 6), 1337);
    map.insert(Key(28, 3), 1337);
    map.insert(Key(4, 4), 1337);
    map.insert(Key(6, 6), 1337);
    map.insert(Key(10, 2), 1337);
    map.insert(Key(20, 3), 1337);
    map.insert(Key(11, 1), 1337);
    map.insert(Key(27, 5), 1337);
    map.insert(Key(28, 5), 1337);
    map.insert(Key(19, 7), 1337);
    map.insert(Key(26, 1), 1337);
    map.insert(Key(25, 8), 1337);
    map.insert(Key(26, 10), 1337);
    map.insert(Key(9, 6), 1337);
    map.insert(Key(29, 14), 1337);
    map.insert(Key(9, 2), 1337);
    map.insert(Key(12, 7), 1337);
    map.insert(Key(21, 4), 1337);
    map.insert(Key(9, 4), 1337);
    map.insert(Key(26, 9), 1337);
    map.insert(Key(14, 4), 1337);
    map.insert(Key(22, 7), 1337);
    map.insert(Key(3, 0), 1337);

    // The following line will hang the program...
    let res = rkyv::to_bytes::<rkyv::rancor::Error>(&map).unwrap();
    
    // ...so this line will never be executed.
    println!("{:?}", res);
}
```

**Comment by djkoloski** [comment]
Fixed by 9cfaa06f21fd8373589d44f63b0ae2de8cf19ee0 and released in rkyv 0.8.1.

**Comment by pkwagner** [comment]
Amazing, thanks a lot!
